File: EIDSSAVR/eidss.remoting/RemoteEventManager.cs
using System.Collections;
using Microsoft.VisualBasic;
using System.Diagnostics;
using System;
using System.Runtime.InteropServices;
using System.Reflection;

namespace eidss.remoting
{
	public delegate void ShowEventDetailHandler(object EventID);
	[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]public interface IRemoteServer
		{
		void ShowEventDetail(object EventId);
		object MainForm{
			get;
			set;
		}
		
	}
	
	public class RemoteEventManager : MarshalByRefObject, IRemoteServer
	{
		
		private static System.Timers.Timer m_Timer;
		private static Queue m_EventsQueue = new Queue();
		public const string Uri = "EIDSS_RemotingServer";
		
		public RemoteEventManager()
		{
			Console.WriteLine("RemoteEventManager activated");
		}
		private static RemoteEventManager m_Singleton = new RemoteEventManager();
		public static RemoteEventManager Singleton
		{
			get
			{
				return m_Singleton;
			}
		}
		
		
		public void ShowEventDetail(object EventID)
		{
			if (m_Timer == null)
			{
				m_Timer = new System.Timers.Timer();
				m_Timer.Elapsed += new System.Timers.ElapsedEventHandler(Timer_Elapsed);
			}
			m_EventsQueue.Enqueue(EventID);
			m_Timer.Start();
		}
		
		private void Timer_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
		{
			m_Timer.Stop();
			object EventID = null;
			if (m_EventsQueue.Count > 0)
			{
				EventID = m_EventsQueue.Dequeue();
			}
			if (EventID != null)
			{
				System.Reflection.MethodInfo mi;
				mi = Singleton.MainForm.GetType().GetMethod("ShowEventDetail");
				if (mi != null)
				{
					Delegate d = Delegate.CreateDelegate(typeof(ShowEventDetailHandler), Singleton.MainForm, mi);
					//We should call ShowEventDetail in the thread of main application,
					//to do this use Invoke method of main application form
					((System.Windows.Forms.Control) Singleton.MainForm).Invoke(d, new object[] {EventID});
				}
			}
			if (m_EventsQueue.Count > 0)
			{
				m_Timer.Start();
			}
		}
		
		
		private object m_MainForm;
		public object MainForm
		{
			get
			{
				return m_MainForm;
			}
			set
			{
				m_MainForm = value;
			}
		}
		
		public static string RemoteURL
		{
			get
			{
				return string.Format("tcp://localhost:{0}/{1}", TcpPort, Uri);
			}
		}
		private static int m_port = 0;
		public static int TcpPort
		{
			get
			{
				if (m_port == 0)
				{
					m_port = System.Convert.ToInt32(bv.common.Configuration.BaseSettings.TcpPort);
				}
				return m_port;
			}
		}
	}
	
	
}

File: EIDSSAVR/vb/EIDSS/EIDSS_Remoting/RemoteEventManager.vb
Imports System.Runtime.InteropServices
Imports System.Reflection
Imports bv.common.Configuration

Public Delegate Sub ShowEventDetailHandler(ByVal EventID As Object)
<InterfaceType(ComInterfaceType.InterfaceIsIUnknown)> _
Public Interface IRemoteServer
    Sub ShowEventDetail(ByVal EventId As Object)
    Property MainForm() As Object

End Interface

Public Class RemoteEventManager
    Inherits MarshalByRefObject
    Implements IRemoteServer
    Private Shared m_Timer As System.Timers.Timer
    Private Shared m_EventsQueue As New Queue
    Public Const Uri As String = "EIDSS_RemotingServer"

    Public Sub New()
        Console.WriteLine("RemoteEventManager activated")
    End Sub
    Private Shared m_Singleton As New RemoteEventManager
    Public Shared ReadOnly Property Singleton() As RemoteEventManager
        Get
            Return m_Singleton
        End Get
    End Property


    Public Sub ShowEventDetail(ByVal EventID As Object) Implements IRemoteServer.ShowEventDetail
        If m_Timer Is Nothing Then
            m_Timer = New System.Timers.Timer
            AddHandler m_Timer.Elapsed, AddressOf Timer_Elapsed
        End If
        m_EventsQueue.Enqueue(EventID)
        m_Timer.Start()
    End Sub

    Private Sub Timer_Elapsed(ByVal sender As Object, ByVal e As System.Timers.ElapsedEventArgs)
        m_Timer.Stop()
        Dim EventID As Object = Nothing
        If m_EventsQueue.Count > 0 Then
            EventID = m_EventsQueue.Dequeue()
        End If
        If Not EventID Is Nothing Then
            Dim mi As Reflection.MethodInfo
            mi = Singleton.MainForm.GetType().GetMethod("ShowEventDetail")
            If Not mi Is Nothing Then
                Dim d As [Delegate] = _
                    [Delegate].CreateDelegate(GetType(ShowEventDetailHandler), Singleton.MainForm, mi)
                'We should call ShowEventDetail in the thread of main application, 
                'to do this use Invoke method of main application form
                CType(Singleton.MainForm, System.Windows.Forms.Control).Invoke(d, New Object() {EventID})
            End If
        End If
        If m_EventsQueue.Count > 0 Then
            m_Timer.Start()
        End If
    End Sub


    Private m_MainForm As Object
    Public Property MainForm() As Object Implements IRemoteServer.MainForm
        Get
            Return m_MainForm
        End Get
        Set(ByVal value As Object)
            m_MainForm = value
        End Set
    End Property

    Public Shared ReadOnly Property RemoteURL() As String
        Get
            Return String.Format("tcp://localhost:{0}/{1}", TcpPort, Uri)
        End Get
    End Property
    Private Shared m_port As Integer = 0
    Private Shared m_defaultPort As Integer = 4001
    Public Shared ReadOnly Property TcpPort() As Integer
        Get
            If m_port = 0 Then
                m_port = BaseSettings.TcpPort
            End If
            Return m_port
        End Get
    End Property
End Class

